prusti-encoder/src/encoders/mir_impure.rs

```diff
@@ -269,9 +162,34 @@ impl<'tcx: 'vir, 'vir> PureFuncAppEnc<'tcx, 'vir> for EncVisitor<'tcx, 'vir, '_>
     ) -> vir::ExprGen<'vir, Self::Curr, Self::Next> {
         self.encode_operand_snap(operand)
     }
+
+    fn monomorphize(&self) -> bool {
+        self.monomorphize
+    }
+}
+
+struct EncodePlaceResult<'vir> {
+    expr: vir::Expr<'vir>,
+
+    /// Statements to undo the impure casts that were made to access the place.
```

Comment:
For example in:
```
fn copy_pair(_1: Pair<T, bool>) -> Pair<T, bool> {
    debug pair => _1;
    let mut _0: Pair<T, bool>;
    let mut _2: T;
    let mut _3: bool;

    bb0: {
        StorageLive(_2);
        _2 = move (_1.0: T);
        StorageLive(_3);
        _3 = (_1.1: bool);
        _0 = Pair::<T, bool> { first: move _2, second: move _3 };
        StorageDead(_3);
        drop(_2) -> [return: bb1, unwind: bb3];
    }

    bb1: {
        StorageDead(_2);
        drop(_1) -> [return: bb2, unwind: bb4];
    }
...
}
```

The line `_3 = (_1.1)` first requires an impure cast of `_1.1` from `T` to `bool`, but since it's a copy the reverse impure cast must be made after the copy to restore permissions for `_1.1` to how they were before.